src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -2201,27 +2202,36 @@ public ArrayType makeArrayType(Type t) {
      * this method could yield surprising answers when invoked on arrays. For example when
      * invoked with t being byte [] and sym being t.sym itself, asSuper would answer null.
      *
-     * @param t a type
-     * @param sym a symbol
-     */
-    public Type asSuper(Type t, Symbol sym) {
-        return asSuper(t, sym, false);
-    }
-
-    /**
-     * Return the (most specific) base type of t that starts with the
-     * given symbol.  If none exists, return null.
+     * Further caveats in Valhalla: There are two "hazards" we need to watch out for when using
+     * this method.
      *
-     * Caveat Emptor: Since javac represents the class of all arrays with a singleton
-     * symbol Symtab.arrayClass, which by being a singleton cannot hold any discriminant,
-     * this method could yield surprising answers when invoked on arrays. For example when
-     * invoked with t being byte [] and sym being t.sym itself, asSuper would answer null.
+     * 1. Since Foo.ref and Foo.val share the same symbol, that of Foo.class, a call to
+     *    asSuper(Foo.ref.type, Foo.val.type.tsym) would return non-null. This is NOT correct
```

Comment:
I don't think I agree with the spirit of this comment :-)

There is only one symbol, so Foo.val.type.tsym == Foo. So:

asSuper(Foo.ref.type, Foo.val.type.tsym) ==
asSuper(Foo.ref.type, Foo) ==
true

This is not different from when you have:

asSuper(ListOfString.type, ListOfInteger.type.tsym) != null

I think the real issue here, is that we can't rely on "asSuper" for code generation, because, really, this routine isn't sharp enough. In the case of Gen::visitTypeCast, the issue is that we have to decide whether a checkcast should be emitted or not; we typically do that by comparing symbols. That strategy here is simply insufficient - as the symbols are the same - just the types will be different. This will be, again, the case when we will face a cast from List to List<int> - these two types will have same symbol, but different representation in the classfile, so a checkcast will be needed there.

Reply:
> I don't think I agree with the spirit of this comment :-)
> 

Indeed, it is stronger than it ought to be. It should say it may not be always what is desired at call site.